### ISSUE TITLE
More canvas for tool search results

### ISSUE THREAD

**Issue by nekrut** [galaxyproject/galaxy#8772]
Galaxy contains many tools (which is fantastic). This however makes it difficult to find tools and the current tool search is not that helpful (also see #3276). 

# The problem

Here is an example, suppose we are searching for tools that deal with `fastq` format. So one would type `fastq` in the search box on top of the tool panel and will get this:

![fastq-search](https://user-images.githubusercontent.com/4291636/66417993-2e8ceb00-e9cf-11e9-99c0-9dc8775c20cd.png)

As one can see the results are listed according to tool categories and **not** by ranks given by the search engine:

`GET https://usegalaxy.org/api/tools?q=fastq`

```
0: "toolshed.g2.bx.psu.edu/repos/devteam/fastqc/fastqc/0.52"
1: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_combiner/fastq_combiner/1.1.1"
2: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_combiner/fastq_combiner/1.0.1"
3: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_groomer/fastq_groomer/1.1.1"
4: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_groomer/fastq_groomer/1.0.4"
5: "toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/1.16.3"
6: "toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/1.16.5"
7: "toolshed.g2.bx.psu.edu/repos/iuc/mothur_make_fastq/mothur_make_fastq/1.39.5.0"
8: "toolshed.g2.bx.psu.edu/repos/iuc/mothur_make_fastq/mothur_make_fastq/1.36.1.0"
9: "toolshed.g2.bx.psu.edu/repos/iuc/mothur_fastq_info/mothur_fastq_info/1.39.5.0"
10: "toolshed.g2.bx.psu.edu/repos/iuc/mothur_fastq_info/mothur_fastq_info/1.36.1.0"
11: "toolshed.g2.bx.psu.edu/repos/jjohnson/mothur_toolsuite/mothur_make_fastq/1.20.0"
12: "toolshed.g2.bx.psu.edu/repos/jjohnson/mothur_toolsuite/mothur_fastq_info/1.25.0"
13: "toolshed.g2.bx.psu.edu/repos/devteam/picard/picard_SamToFastq/1.56.1"
14: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_masker_by_quality/fastq_masker_by_quality/1.0.0"
15: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_manipulation/fastq_manipulation/1.0.1"
16: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_interlacer/fastq_paired_end_interlacer/1.2.0"
17: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_splitter/fastq_paired_end_splitter/1.0.0"
18: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_joiner/fastq_paired_end_joiner/2.0.1.0"
19: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_joiner/fastq_paired_end_joiner/2.0.1.1"
20: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_filter/fastq_filter/1.1.1"
21: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_trimmer/fastq_trimmer/1.0.0"
22: "toolshed.g2.bx.psu.edu/repos/lparsons/fastq_join/fastq_join/1.1.2-806.1"
23: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_trimmer_by_quality/fastq_quality_trimmer/1.0.0"
24: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_interlacer/fastq_paired_end_interlacer/1.2.0.1"
25: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_filter/fastq_filter/1.0.0"
26: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_deinterlacer/fastq_paired_end_deinterlacer/1.1.1"
27: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_deinterlacer/fastq_paired_end_deinterlacer/1.1.2"
28: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_masker_by_quality/fastq_masker_by_quality/1.1.1"
29: "toolshed.g2.bx.psu.edu/repos/devteam/fastq_manipulation/fastq_manipulation/1.1.1"
```

in fact, the top result, `fastqc` **is not displayed at all**! Thus the search is not useful.

# A potential solution
(as suggested by @martenson)

![Screen Shot 2019-10-08 at 1 43 33 PM](https://user-images.githubusercontent.com/4291636/66419214-ac51f600-e9d1-11e9-99ab-3e60d1470086.png)

Here the tool panel does not change when you type the search term. Instead a modal appears (can also simply use the center pane) listing tools ranked by search engine (not by tool categories). You can imagine also adding some actions for each displayed tool such as "add to favorites" etc. 

The search is based on index created from title, description, and help section of the tool XML. If we develop a tagging system for tool, this can also be added to the index.

**Comment by hexylena** [comment]
Sounds a lot like the [tool search webhook](https://github.com/galaxyproject/galaxy/pull/3460) that was written by @anuprulez

**Comment by jxtx** [comment]
Rather than another canvas, could just change the sorting in the existing tool panel to be ranked.

**Comment by martenson** [comment]
@jxtx copy from private channel 

1. I think we will need to display more information in the future like tags/related tools/data types etc. for which more space is needed since the tool panel is really narrow
2. I’d like to accommodate two types of users. The ‘pros’ that know the tool panel like the back of their hand and navigate visually, and the new users that are lost in the hundreds of tools and dozens of sections. All of that without extra interaction with filters/ordering - I believe people don’t tend to interact with search results other ways than by scrolling

edit: that said implementing ordered tool panel within the existing space is a fine step

**Comment by jxtx** [comment]
Yes, I agree, just thinking it would make sense to take it incrementally.

**Comment by martenson** [comment]
@jxtx would you change the default to 'ranked'?

**Comment by jxtx** [comment]
That probably makes sense. There is also probably a reason we didn't do that in the first place. Let's see how it looks.

**Comment by scholtalbers** [comment]
Going through a tutorial and being on usegalaxy.eu with a-lot-of-tools the inability to search for mentioned tools is a major pain. Also, I'm not sure if the `pro` user is accommodated well with the standard tool panel? The pro will probably know the tool names and can easily search for them. The same as I want to go to a website, I type the name and let the browser/google do the rest. I'm not going through my bookmarks to find it...  - in addition, with different Galaxy flavors, tool panels are likely not consistent across them.
+1 for just displaying the ranked results in the tool panel.

**Comment by dannon** [comment]
The search ordering is being addressed in #9579

**Comment by jmchilton** [comment]
I think a lot of this was addressed by #9579 - I know tool search is still imperfect but I'm going to close this and let people reopen new issues based on the new 21.01 as a basis for discussion. If I'm wildly off-base and this issue should be reopened please just re-open and disregard this message.
